dot/rpc/modules/chain_test.go

```diff
@@ -15,3 +15,448 @@
+	i := make([]interface{}, 1)
+	i[0] = "a"
+
+	mockBlockAPI := new(apimocks.BlockAPI)
+	mockBlockAPI.On("BestBlockHash").Return(testHash, nil)
+	mockBlockAPI.On("GetBlockHash", mock.AnythingOfType("*big.Int")).Return(testHash, nil)
+
+	mockBlockAPIErr := new(apimocks.BlockAPI)
+	mockBlockAPIErr.On("BestBlockHash").Return(testHash, nil)
+	mockBlockAPIErr.On("GetBlockHash", mock.AnythingOfType("*big.Int")).Return(nil, errors.New("GetBlockHash Error"))
+
+	var res ChainHashResponse
+	type fields struct {
+		blockAPI BlockAPI
+	}
+	type args struct {
+		r   *http.Request
+		req *ChainBlockNumberRequest
+		res *ChainHashResponse
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "GetBlockHash nil req OK",
+			fields: fields{
+				mockBlockAPI,
+			},
+			args: args{
+				req: &ChainBlockNumberRequest{},
+				res: &res,
+			},
+		},
+		{
+			name: "GetBlockHash string req OK",
+			fields: fields{
+				mockBlockAPI,
+			},
+			args: args{
+				req: &ChainBlockNumberRequest{"21"},
+				res: &res,
+			},
+		},
+		{
+			name: "GetBlockHash float req OK",
+			fields: fields{
+				mockBlockAPI,
+			},
+			args: args{
+				req: &ChainBlockNumberRequest{float64(21)},
+				res: &res,
+			},
+		},
+		{
+			name: "GetBlockHash unknown request number",
+			fields: fields{
+				mockBlockAPI,
+			},
+			args: args{
+				req: &ChainBlockNumberRequest{uintptr(1)},
+				res: &res,
+			},
+			wantErr: true,
+		},
+		{
+			name: "GetBlockHash string slice req err",
+			fields: fields{
+				mockBlockAPI,
+			},
+			args: args{
+				req: &ChainBlockNumberRequest{i},
+				res: &res,
+			},
+			wantErr: true,
+		},
+		{
+			name: "GetBlockHash string req Err",
+			fields: fields{
+				mockBlockAPIErr,
+			},
+			args: args{
+				req: &ChainBlockNumberRequest{"21"},
+				res: &res,
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cm := &ChainModule{
+				blockAPI: tt.fields.blockAPI,
+			}
+			if err := cm.GetBlockHash(tt.args.r, tt.args.req, tt.args.res); (err != nil) != tt.wantErr {
+				t.Errorf("GetBlockHash() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func TestChainModule_GetFinalizedHead(t *testing.T) {
+	testHash := common.NewHash([]byte{0x01, 0x02})
+
+	mockBlockAPI := new(apimocks.BlockAPI)
+	mockBlockAPI.On("GetHighestFinalisedHash").Return(testHash, nil)
+
+	mockBlockAPIErr := new(apimocks.BlockAPI)
+	mockBlockAPIErr.On("GetHighestFinalisedHash").Return(nil, errors.New("GetHighestFinalisedHash Error"))
+
+	var res ChainHashResponse
+	type fields struct {
+		blockAPI BlockAPI
+	}
+	type args struct {
+		r   *http.Request
+		req *EmptyRequest
+		res *ChainHashResponse
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "GetHighestFinalisedHash OK",
+			fields: fields{
+				mockBlockAPI,
+			},
+			args: args{
+				req: &EmptyRequest{},
+				res: &res,
```

Comment:
That will likely mess up if we run each subtest in parallel later, since I would assume they all write to the same response in parallel?

Instead define it in the subtest.

As a general rule, it's usually better to define things in the subtest (in the for loop) scope instead of the parent scope, especially when it comes to pointy pointers/interfaces.

Reply:
To clarify, should I just define the response in the subtest, or should I be defining the test cases in their entirety in the subtest?

Reply:
I'm not sure anymore since the code changed when I pulled it haha! I think it was about `res` being shared among subtests and written to.

Also, you can also write your test cases without `args` and `fields` structs and have everything at the same level of indentation, having tried both, I find it cleaner. But up to you!